ACPs/99-validatorsetmanager-contract/README.md

```diff
@@ -0,0 +1,264 @@
+ * @param weight The new weight of the node on the Subnet
+ */
+function setSubnetValidatorWeight(
+    bytes32 nodeID,
+    uint64 weight
+) external onlyOwner;
+
+/**
+ * Called to update the Subnet validator set upon receiving a notification Warp message from the P-Chain
+ * (after the successful addition or removal of a validator)
+ * @param messageIndex The index of the Warp message from the P-Chain
+ */
+function updateSubnetValidatorSet(
+    uint32 messageIndex
+) external;
+```
+
+#### About `ExitValidatorSetTx`
+
+This transaction allows a validator to exit a Subnet validator set directly from the P-Chain to prevent the Subnet from locking validators into participating in consensus indefinitely. Therefore it is not meant to be called in the `ValidatorSetManager` contract.
+
+### Information Needed from the P-Chain
+
+Upon successful addition, weight update, and removal of a validator from the Subnet, the P-Chain will send a notification in the form of a Warp message. The `updateSubnetValidatorSet` function is responsible for deserializing this message and updating the contract state accordingly.
+
+To update the validator set, the P-Chain needs to send a notification with the following payload:
+
+```
++-----------+----------+-----------+
+|  subnetID : [32]byte |  32 bytes |
++-----------+----------+-----------+
+|    nodeID : [32]byte |  32 bytes |
++-----------+----------+-----------+
+| messageID : [32]byte |  32 bytes |
++-----------+----------+-----------+
+|     nonce :   uint64 |   8 bytes |
++-----------+----------+-----------+
+|    weight :   uint64 |   8 bytes |
++-----------+----------+-----------+
+| timestamp :   uint64 |   8 bytes |
++-----------+----------+-----------+
+|    uptime : [64]byte |  64 bytes |
++-----------+----------+-----------+
+                       | 184 bytes |
+                       +-----------+
+```
+
+- The Subnet ID
+- The NodeID
+- The `MessageID` corresponding to the Warp message in the `RegisterSubnetValidatorTx` that added the validator (SHA256)
+- The `nonce` corresponding to this update
+- The weight of the validator (if `0`, this means that the validation is over)
+- The exact timestamp at which the weight was updated (if the nonce is `0`, this also means the beginning of the `validation`)
+- The uptime observed during the previous validation period (only useful in case a validator weight update)
+
+Note that the timestamp can also be used to update the `endTime` of the previous validation period.
+
+**Note:** Once this discussion/ACP is in an advanced stage, an update to ACP-77 can be made to specify the payload of P-Chain notifications.
+
+### Reference Architecture for Custom “Security Modules”
+
+The `ValidatorSetManager` contract will be ownable (e.g. using [OpenZeppelin Ownable](https://docs.openzeppelin.com/contracts/5.x/api/access#Ownable)). The ownership can be transferred to a smart contract, the “Security Module”, that will implement the security rules for the Subnet.
```

Comment:
Given how crucial this module will be, it may be best practice to enforce a two-step mechanism

Reply:
Could you expand on what those 2 steps would be?

Reply:
OZ's [Ownable2Step](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/access/Ownable2Step.sol) for example. the new owner needs to accept the ownership transfer before it becomes effective. though this might be harder to handle if the owner is not a smart account but a regular contract.

Reply:
The likely case for using a contract instead of smart account would be PoS validator sets, in which case the contract interface should likely have an `acceptOwnership` function, and should already be taking in the `ValidatorManager` address. 

Mostly a question of not introducing extra access controls for making `acceptOwnership` call, thinking if this would be done in the constructor